diff --git a/sdk/objc/components/capturer/RTCCameraVideoCapturer.m b/sdk/objc/components/capturer/RTCCameraVideoCapturer.m
index 744edf6..f83c03e 100644
--- a/sdk/objc/components/capturer/RTCCameraVideoCapturer.m
+++ b/sdk/objc/components/capturer/RTCCameraVideoCapturer.m
@@ -41,6 +41,7 @@
   RTCVideoRotation _rotation;
 #if TARGET_OS_IPHONE
   UIDeviceOrientation _orientation;
+  BOOL _generatingOrientationNotifications;
 #endif
 }
 
@@ -158,7 +159,12 @@
                       RTCLogInfo("startCaptureWithDevice %@ @ %ld fps", format, (long)fps);
 
 #if TARGET_OS_IPHONE
-                      [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
+                      dispatch_async(dispatch_get_main_queue(), ^{
+                        if (!self->_generatingOrientationNotifications) {
+                          [[UIDevice currentDevice] beginGeneratingDeviceOrientationNotifications];
+                          self->_generatingOrientationNotifications = YES;
+                        }
+                      });
 #endif
 
                       self.currentDevice = device;
@@ -200,7 +206,12 @@
                       [self.captureSession stopRunning];
 
 #if TARGET_OS_IPHONE
-                      [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];
+                      dispatch_async(dispatch_get_main_queue(), ^{
+                        if (self->_generatingOrientationNotifications) {
+                          [[UIDevice currentDevice] endGeneratingDeviceOrientationNotifications];
+                          self->_generatingOrientationNotifications = NO;
+                        }
+                      });
 #endif
                       self.isRunning = NO;
                       if (completionHandler) {
